presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1321,6 +1330,49 @@ protected Scope visitTable(Table table, Optional<Scope> scope)
             return createAndAssignScope(table, scope, fields.build());
         }
 
+        private Optional<TableHandle> getTableHandle(TableColumnMetadata tableColumnsMetadata, Table table, QualifiedObjectName name, Optional<Scope> scope)
+        {
+            // Process table version AS OF expression
+            if (table.getTableVersionExpression().isPresent()) {
+                return processTableVersion(table, name, scope);
+            }
+            else {
+                return tableColumnsMetadata.getTableHandle();
+            }
+        }
+        private Optional<TableHandle> processTableVersion(Table table, QualifiedObjectName name, Optional<Scope> scope)
+        {
+            Expression asOfExpr = table.getTableVersionExpression().get().getAsOfExpression();
+            TableVersionExpression.TableVersionType tableVersionType = table.getTableVersionExpression().get().getTableVersionType();
+            ExpressionAnalysis expressionAnalysis = analyzeExpression(asOfExpr, scope.get());
+            analysis.recordSubqueries(table, expressionAnalysis);
+            Type asOfExprType = expressionAnalysis.getType(asOfExpr);
+            if (asOfExprType == UNKNOWN) {
+                throw new PrestoException(INVALID_ARGUMENTS, format("Table version AS OF expression cannot be NULL for %s", name.toString()));
+            }
+            Object evalAsOfExpr = evaluateConstantExpression(asOfExpr, asOfExprType, metadata, session, analysis.getParameters());
+            if (tableVersionType == TIMESTAMP) {
+                if (!(asOfExprType instanceof TimestampWithTimeZoneType)) {
+                    throw new SemanticException(TYPE_MISMATCH, asOfExpr,
+                            "Type %s is invalid. Supported table version AS OF expression type is Timestamp with Time Zone.",
+                            asOfExprType.getDisplayName());
+                }
+            }
```

Comment:
Is there a possibility that we also support expression of type TimestampType, and internally change it to TimestampWithTimeZoneType by some explicitly defined semantics? What's your opinion?

Reply:
@hantangwangd , I had discussed this issue with @tdcmeehan. If we use session's time zone, query may give different results for different time zone regions. It could be misleading to end users, since TIMESTAMP syntax will match closest snapshot id in the catalog table. User will need to know system's time zone used in table creation. we can use TimestampWithTimeZone (default time zone) for DATE and TIMESTAMP cases and document it. Or create a new issue and fix after first release (based on customer feedback). I will update more details tomorrow.

Reply:
Sounds good. Agree with you.